oak/server/oak_node.cc

```diff
@@ -186,123 +141,119 @@ OakNode::OakNode(const std::string& node_id, const std::string& module)
-    }
-    WriteMemory(env, offset, module_data_input_.cbegin(), module_data_input_.cbegin() + size);
-    results[0].set_i32(size);
-    module_data_input_.remove_prefix(size);
+    ChannelId channel_id = args[0].get_i64();
+    uint32_t offset = args[1].get_i32();
+    uint32_t size = args[2].get_i32();
+
+    std::unique_ptr<Channel>& channel = channels_[channel_id];
+
+    absl::Span<const char> data = channel->Read(size);
+    WriteMemory(env, offset, data);
+    results[0].set_i32(data.size());
 
     return wabt::interp::Result::Ok;
   };
 }
 
-wabt::interp::HostFunc::Callback OakNode::OakWrite(wabt::interp::Environment* env) {
+wabt::interp::HostFunc::Callback OakNode::OakWriteChannel(wabt::interp::Environment* env) {
   return [this, env](const wabt::interp::HostFunc* func, const wabt::interp::FuncSignature* sig,
                      const wabt::interp::TypedValues& args, wabt::interp::TypedValues& results) {
-    LOG(INFO) << "Called host function: " << func->module_name << "." << func->field_name;
-    for (auto const& arg : args) {
-      LOG(INFO) << "Arg: " << wabt::interp::TypedValueToString(arg);
-    }
-    uint32_t offset = args[0].get_i32();
-    uint32_t size = args[1].get_i32();
+    LogHostFunctionCall(func, args);
+
+    ChannelId channel_id = args[0].get_i64();
+    uint32_t offset = args[1].get_i32();
+    uint32_t size = args[2].get_i32();
 
-    absl::Span<const char> memory = ReadMemory(env, offset, size);
-    module_data_output_->insert(module_data_output_->end(), memory.cbegin(), memory.cend());
-    results[0].set_i32(size);
+    std::unique_ptr<Channel>& channel = channels_[channel_id];
+
+    absl::Span<const char> data = ReadMemory(env, offset, size);
+    channel->Write(data);
+    results[0].set_i32(data.size());
 
     return wabt::interp::Result::Ok;
   };
 }
 
 grpc::Status OakNode::ProcessModuleInvocation(grpc::GenericServerContext* context,
-                                              const std::vector<uint8_t>& request_data,
-                                              std::vector<uint8_t>* response_data) {
+                                              const std::vector<char>& request_data,
+                                              std::vector<char>* response_data) {
   LOG(INFO) << "Handling gRPC call: " << context->method();
   server_context_ = context;
 
-  absl::MutexLock lock(&module_data_mutex_);
-  module_data_input_ = absl::Span<const uint8_t>(request_data);
-  module_data_output_ = response_data;
+  // Create the gRPC channel, used by the module to perform basic input and output.
+  std::unique_ptr<Channel> grpc_channel =
+      absl::make_unique<BufferChannel>(request_data, response_data);
+  ChannelId grpc_channel_id = NewChannelId();
+  LOG(INFO) << "Created gRPC channel: " << grpc_channel_id;
+  channels_[grpc_channel_id] = std::move(grpc_channel);
+
+  // Store a copy of the gRPC method name so that we can refer to it via a Span when the Module
```

Comment:
Maybe it would be simpler to just store the method name in the channel/stream, even if it's not used by all methods.

Reply:
Yes I think for this particular use case it would be simpler if the Channel owned the data, but for other use cases it's nice for it to just use a lightweight reference (span). Is there any reasonable way of supporting both modes in the same class?

Reply:
I can't think of anything offhand.